### PR TITLE
A fog-openstack with a more "version less" approach

### PR DESCRIPTION
**Pull request by gildub**
Changes necessary to use fog-openstack supporting "version less" endpoint (including identity) and which avoid tempering with URL path.

Tested: Cloud/Infra on Gaprindashvili and master

Requires https://github.com/ManageIQ/manageiq-gems-pending/pull/386

### CODE REVIEW

**Comment by gildub** [comment]
@aufi, @mansam, @sseago, @Ladas feedback please.

**Comment by sseago** [comment]
Looks good to me. Will the fog-openstack dependency be updated in the gemspec as part of this PR once the dependent fog-openstack changes are merged and included in a release?

File: spec/legacy/openstack_handle/handle_spec.rb [hunk 1]
```diff
@@ -49,62 +49,60 @@
...
+        :openstack_project_domain_id => nil,
+        :openstack_user_domain_id    => nil,
+        :openstack_region            => nil,
+        :connection_options          => {:ssl_verify_peer => false}
+      ).once do |_, _, address|
         expect(address).to eq(auth_url)
         fog
       end
-      expect(handle.connect(:tenant_name => "admin")).to eq(fog)
+      expect(handle.connect(:openstack_project_name => "admin")).to eq(fog)
```
Comment (mansam) [inline spec/legacy/openstack_handle/handle_spec.rb]: I might not be tracing the logic correctly, but it looks like passing `openstack_project_name` here doesn't quite make sense. It looks like `connect` sees the `tenant_name` field is empty, but the `discover_tenants` option defaults to true so it sets `tenant = default_tenant_name`. When it gets to the conditional starting on line 166, it sees `tenant` is set so it uses that value, overwriting the value of `openstack_project_name` you passed in.

**Comment by gildub** [comment]
@aufi, @mansam,

While testing against gaprindashvili for cloud provider with v2 authentication, I get the following error:

```ruby
ERROR -- : MIQ(OpenstackHandle::ComputeDelegate#handled_list) Unable to obtain collection: 'services' in service: 'nova' using project scope: '' in provider: '192.0.2.6'. Message=undefined method `url' for OpenstackHandle::Handle:Class
ERROR -- : MIQ(OpenstackHandle::ComputeDelegate#handled_list) /home/gildub/manageiq-providers-openstack/lib/manageiq/providers/openstack/legacy/openstack_handle/identity_delegate.rb:38:in `block in visible_tenants
_v2'
``` 

Maybe I missed something but I can't find the any `url` method for `Handle` class, parent class or related module and it's not in delegated class either. Any clue?

**Comment by mansam** [comment]
@gildub the `self.url` method of the `Handle` class was on line 107 of `lib/manageiq/providers/openstack/legacy/openstack_handle/handle.rb`, but you appear to have removed it  or consolidated it into `self.auth_url` as part of this PR. Hope that helps!

**Comment by gildub** [comment]
@mansam, Oh my, of course! Thanks.

**Comment by mansam** [comment]
I've been testing this and the Fog changes against the Liberty environments we use for recording VCRs. Most things appear to work, but I see problems with malformed URLs when trying to list visible projects on Keystone v3. It appears that the keystone endpoints on the failing environment are versioned.

```
     # Excon::Error::NotFound:
     #   Expected([200]) <=> Actual(404 Not Found)
     #   excon.error.response
     #     :body          => "{\"error\": {\"message\": \"The resource could not be found.\", \"code\": 404, \"title\": \"Not Found\"}}"
     #     :cookies       => [
     #     ]
     #     :headers       => {
     #       "Connection"             => "close"
     #       "Content-Length"         => "93"
     #       "Content-Type"           => "application/json"
     #       "Date"                   => "Tue, 28 Aug 2018 15:35:31 GMT"
     #       "Server"                 => "Apache/2.4.6 (Red Hat Enterprise Linux)"
     #       "Vary"                   => "X-Auth-Token"
     #       "x-openstack-request-id" => ""
     #     }
     #     :host          => "
     #     :local_address => ""
     #     :local_port    => 37474
     #     :path          => "/v3/v3/users/0e379b058aea4a7ba72d13cddf21c240/projects"
     #     :port          => 5000
     #     :reason_phrase => "Not Found"
     #     :remote_ip     => ""
     #     :status        => 404
     #     :status_line   => "HTTP/1.1 404 Not Found\r\n"
     #   ./lib/manageiq/providers/openstack/legacy/openstack_handle/identity_delegate.rb:28:in `visible_tenants_v3'
```
A couple of other things seem to break as well which I'm still investigating and will update when I have more details.

**Comment by gildub** [comment]
@mansam, thank you very much for the feedback. 
https://github.com/fog/fog-openstack/pull/418 should address the case where the endpoint is 'versioned'.

**Comment by gildub** [comment]
@mansam, actually another patch is needed to fully cover the versionned endpoint:
https://github.com/fog/fog-openstack/pull/421

**Comment by gildub** [comment]
@mansam, @aufi, @Ladas 

Tracking down the Rspec failure with ./spec/models/manageiq/providers/openstack/cloud_manager/refresher_rhos_liberty_keystone_v3_spec.rb
It seems we need to re-run VCR for because a "GET http://11.22.33.44:5000/v3/projects" is triggered by line:
https://github.com/ManageIQ/manageiq-providers-openstack/blob/master/lib/manageiq/providers/openstack/legacy/openstack_handle/identity_delegate.rb#L28
but missing from the cassette.

**Comment by gildub** [comment]
- spec/models/manageiq/providers/openstack/cloud_manager/refresher_rhos_liberty_spec.rb
- spec/models/manageiq/providers/openstack/infra_manager/refresher_rhos_juno_spec.rb

Are both had an issue is related with an unscoped token in identity v2.0. 
There was also an issue with baremetal service
Those have been addressed and will be into fog-openstack 0.2.4.

**Comment by mansam** [comment]
Hi @gildub, I'm not sure if this PR is ready for testing again or not, but I tried rerecording the VCRs again with the new version of fog. Here's what I'm seeing:

On the Liberty w/ Keystone V3 VCR, I'm still seeing the following failure:

```
1) ManageIQ::Providers::Openstack::CloudManager::Refresher when using an admin account for fast refresh will perform a fast full refresh against RHOS 
     Failure/Error: projects.all(:domain_id => @os_handle.domain_id, :user_id => current_user_id)
     
     Fog::Identity::OpenStack::NotFound:
       Expected([200]) <=> Actual(404 Not Found)
       excon.error.response
         :body          => "{\"error\": {\"message\": \"The resource could not be found.\", \"code\": 404, \"title\": \"Not Found\"}}"
         :cookies       => [
         ]
         :headers       => {
           "Connection"             => "close"
           "Content-Length"         => "93"
           "Content-Type"           => "application/json"
           "Date"                   => "Thu, 13 Sep 2018 20:03:04 GMT"
           "Server"                 => "Apache/2.4.6 (Red Hat Enterprise Linux)"
           "Vary"                   => "X-Auth-Token"
           "x-openstack-request-id" => "req-17774be3-1af7-4257-8934-a59be5a8cffe"
         }
         :host          => "10.8.99.206"
         :local_address => "10.10.122.183"
         :local_port    => 32914
         :path          => "/v3/v3/users/0e379b058aea4a7ba72d13cddf21c240/projects"
         :port          => 5000
         :reason_phrase => "Not Found"
         :remote_ip     => "10.8.99.206"
         :status        => 404
         :status_line   => "HTTP/1.1 404 Not Found\r\n"
     # ./lib/manageiq/providers/openstack/legacy/openstack_handle/identity_delegate.rb:28:in `visible_tenants_v3'

```
For the Liberty w/ V2 specs, I'm seeing the following while trying to rerecord:

```
      Failure/Error: raise MiqException::ServiceNotAvailable
      
      MiqException::ServiceNotAvailable:
        MiqException::ServiceNotAvailable
      # ./lib/manageiq/providers/openstack/legacy/openstack_handle/handle.rb:89:in `rescue in raw_connect'
      # ./lib/manageiq/providers/openstack/legacy/openstack_handle/handle.rb:60:in `raw_connect'
      # ./lib/manageiq/providers/openstack/legacy/openstack_handle/handle.rb:56:in `block in raw_connect_try_ssl'
      # ./lib/manageiq/providers/openstack/legacy/openstack_handle/handle.rb:45:in `try_connection'
      # ./lib/manageiq/providers/openstack/legacy/openstack_handle/handle.rb:53:in `raw_connect_try_ssl'
      # ./lib/manageiq/providers/openstack/legacy/openstack_handle/handle.rb:182:in `connect'
      # ./lib/manageiq/providers/openstack/legacy/openstack_handle/handle.rb:226:in `identity_service'
      # ./lib/manageiq/providers/openstack/legacy/openstack_handle/handle.rb:332:in `tenants'
      # ./lib/manageiq/providers/openstack/legacy/openstack_handle/handle.rb:349:in `accessible_tenants'
      # ./app/models/manageiq/providers/openstack/cloud_manager/refresh_parser.rb:126:in `get_tenants'
      # ./app/models/manageiq/providers/openstack/cloud_manager/refresh_parser.rb:58:in `ems_inv_to_hashes'
      # ./app/models/manageiq/providers/openstack/cloud_manager/refresh_parser.rb:13:in `ems_inv_to_hashes'
      # ./spec/manageiq/app/models/manageiq/providers/base_manager/manager_refresher.rb:20:in `parse_legacy'
      # ./spec/manageiq/app/models/manageiq/providers/base_manager/manager_refresher.rb:65:in `block (2 levels) in parse_targeted_inventory'
      # ./spec/manageiq/app/models/manageiq/providers/base_manager/manager_refresher.rb:65:in `block in parse_targeted_inventory'
      # ./spec/manageiq/app/models/manageiq/providers/base_manager/manager_refresher.rb:61:in `parse_targeted_inventory'
      # ./spec/manageiq/app/models/manageiq/providers/base_manager/refresher.rb:105:in `block in refresh_targets_for_ems'
      # ./spec/manageiq/app/models/manageiq/providers/base_manager/refresher.rb:104:in `refresh_targets_for_ems'
      # ./spec/manageiq/app/models/manageiq/providers/base_manager/refresher.rb:41:in `block (2 levels) in refresh'
      # ./spec/manageiq/app/models/manageiq/providers/base_manager/refresher.rb:41:in `block in refresh'
      # ./spec/manageiq/app/models/manageiq/providers/base_manager/refresher.rb:31:in `each'
      # ./spec/manageiq/app/models/manageiq/providers/base_manager/refresher.rb:31:in `refresh'
      # ./spec/manageiq/app/models/manageiq/providers/base_manager/refresher.rb:11:in `refresh'
      # ./spec/manageiq/app/models/ems_refresh.rb:101:in `block in refresh'
      # ./spec/manageiq/app/models/ems_refresh.rb:100:in `each'
      # ./spec/manageiq/app/models/ems_refresh.rb:100:in `refresh'
      # ./spec/models/manageiq/providers/openstack/cloud_manager/refresher_rhos_liberty_spec.rb:43:in `block (5 levels) in <top (required)>'
      # ./spec/models/manageiq/providers/openstack/cloud_manager/refresh_spec_helpers.rb:56:in `block in with_cassette'
      # ./spec/models/manageiq/providers/openstack/cloud_manager/refresh_spec_helpers.rb:51:in `with_cassette'
      # ./spec/models/manageiq/providers/openstack/cloud_manager/refresher_rhos_liberty_spec.rb:42:in `block (4 levels) in <top (required)>'
      # ./spec/models/manageiq/providers/openstack/cloud_manager/refresher_rhos_liberty_spec.rb:41:in `times'
      # ./spec/models/manageiq/providers/openstack/cloud_manager/refresher_rhos_liberty_spec.rb:41:in `block (3 levels) in <top (required)>'
      # ------------------
      # --- Caused by: ---
      # Fog::OpenStack::Auth::Catalog::ServiceTypeError:
      #   No endpoint match
      #   ./lib/manageiq/providers/openstack/legacy/openstack_handle/handle.rb:84:in `raw_connect'
```

**Comment by gildub** [comment]
@mansam, thanks for trying again. So the case we're hitting right now is when the auth_url doesn't contain 'v3' but the endpoint returned from the catalog for identity does contain 'v3'. I'm working on it!.

**Comment by mansam** [comment]
Hi @gildub, I've been troubleshooting the failing tests and have noticed that when the tests make connections to the identity service with the following parameters:
```ruby
{
 :openstack_auth_url=>"http://<vcr-instance>:5000",
 :openstack_username=>"admin",
 :openstack_api_key=>"<password>",
 :openstack_endpoint_type=>"public",
 :openstack_identity_api_version=>"v2.0",
 :openstack_region=>nil
}
```
then they fail with the following catalog error:
```
Fog::OpenStack::Auth::Catalog::ServiceTypeError: No endpoint match
	from /home/slucidi/Workspace/manageiq/vendor/bundle/ruby/2.3.0/gems/fog-openstack-0.3.6/lib/fog/openstack/auth/catalog.rb:23:in `get_endpoint_url'
	from /home/slucidi/Workspace/manageiq/vendor/bundle/ruby/2.3.0/gems/fog-openstack-0.3.6/lib/fog/openstack/core.rb:212:in `authenticate'
	from /home/slucidi/Workspace/manageiq/vendor/bundle/ruby/2.3.0/gems/fog-openstack-0.3.6/lib/fog/openstack/core.rb:44:in `initialize'
	from /home/slucidi/Workspace/manageiq/vendor/bundle/ruby/2.3.0/gems/fog-core-1.45.0/lib/fog/core/service.rb:115:in `new'
	from /home/slucidi/Workspace/manageiq/vendor/bundle/ruby/2.3.0/gems/fog-core-1.45.0/lib/fog/core/service.rb:115:in `new'
	from /home/slucidi/Workspace/manageiq/vendor/bundle/ruby/2.3.0/gems/fog-openstack-0.3.6/lib/fog/identity/openstack.rb:9:in `new'
	from (irb):2
	from /home/slucidi/Workspace/manageiq/vendor/bundle/ruby/2.3.0/gems/railties-5.0.7/lib/rails/commands/console.rb:65:in `start'
	from /home/slucidi/Workspace/manageiq/vendor/bundle/ruby/2.3.0/gems/railties-5.0.7/lib/rails/commands/console_helper.rb:9:in `start'
	from /home/slucidi/Workspace/manageiq/vendor/bundle/ruby/2.3.0/gems/railties-5.0.7/lib/rails/commands/commands_tasks.rb:78:in `console'
	from /home/slucidi/Workspace/manageiq/vendor/bundle/ruby/2.3.0/gems/railties-5.0.7/lib/rails/commands/commands_tasks.rb:49:in `run_command!'
	from /home/slucidi/Workspace/manageiq/vendor/bundle/ruby/2.3.0/gems/railties-5.0.7/lib/rails/commands.rb:18:in `<top (required)>'
	from bin/rails:4:in `require'
	from bin/rails:4:in `<main>'
```

The `@payload` property on the Catalog object comes back as `[]` which leads directly to the failure of `get_endpoint_url since the endpoint list is empty. However, if the connection parameters are amended to include a value for `:openstack_tenant`, then the catalog is populated and the connection succeeds. I'm not sure if that's due to the tenant being necessary to retrieve the endpoint catalog or if there's something else going on in Fog's internals I haven't found yet, but perhaps you'll have some insight as to what is going on.

**Comment by mansam** [comment]
Looks like the easiest fix is to adjust the conditional [here](https://github.com/fog/fog-openstack/blob/master/lib/fog/openstack/core.rb#L211) to `@openstack_management_url = if token.catalog && !token.catalog.payload.empty?` so that present but empty catalogs fall through to the else. With that change, all the tests pass. That said, I don't know if there are any negative consequences of doing that.

**Comment by gildub** [comment]
@mansam, first off, thank you for tracking that one.

When a tenant is not provided, the token provided is "unscoped" and there is no catalog: 

```bash
$ curl -i \
>   -H "Content-type: application/json" \
>   -d '
> { "auth": {
>     "passwordCredentials": {
>       "username": "admin",
>       "password": "secret"
>     }
> }
> }' \
>    http://${OS_HOST}:5000/v2.0/tokens ; echo

{"access": {"token": {"issued_at": "2018-10-19T00:35:25.000000Z", "expires": "2018-10-19T01:35:25.000000Z", "id": "gAAAAABbySbN3m1lefVC6VJDb_xCS1MZwUtnKLVLjcZ7ItpaHHmoTatJZDXfINKPUzEjqZ1Gpgwk5OH3v6HLIkts6Q1O36oqeLdH0CSwZwNdR_m8QdHVa2IqsKrOmGHfdzLmJVuMj_LUzXnkebWCYiYssOrOhfMayA", "audit_ids": ["vLUxjTJ7TVW_CD9WEiWmLA"]}, "serviceCatalog": [], "user": {"username": "admin", "roles_links": [], "id": "56d5f121c4b94edd90d57703dad4247a", "roles": [], "name": "admin"}, "metadata": {"is_admin": 0, "roles": []}}}
```

This is the expected behavior. 
Meanwhile fog-openstack's authentication should not try to get the endpoint url when the catalog is empty. The legacy code in regard of the unscoped token wasn't very clear to me. Meanwhile I can't see any negative effects with the new code.

**Comment by mansam** [comment]
@gildub Awesome, with that change and manageiq-gems-pending updated to use the same version of Fog, all the tests pass.

**Comment by mansam** [comment]
@gildub I'm running tests with locally rerecorded VCRs. If you're testing with the old VCRs, you'll see failures because the casettes won't line up. If you have difficulty rerecording VCRs, I can push mine.

**Comment by gildub** [comment]
@mansam, I'm going to say yes to your offer ;) . My issue with re-recording VCRs is having the appropriate legacy RHOS environments handy. Do you have them available?

**Comment by mansam** [comment]
@gildub Our team does. I'll email you the environment info, but I'm also happy to just record VCRs for you in the future as well.

**Comment by mansam** [comment]
:fireworks: :fireworks: :fireworks:

**Pull request merged**